examples/actor_critic.py

```diff
@@ -0,0 +1,413 @@
+            self._env.reset(benchmark=FLAGS.benchmark)
+
+            # Project onto the subset of transformations that have
+            # been specified to be used.
+            self._actions = [self._env.action_space.flags.index(f) for f in FLAGS.flags]
+            self.action_space = [self._env.action_space.flags[a] for a in self._actions]
+
+        except:
+            # The program will not terminate until the environment is
+            # closed, not even if there is an exception.
+            self._env.close()
+            raise
+
+    def step(self, action):
+        assert 0 <= action and action < len(self._actions)
+
+        # Update state
+        assert self._steps_taken < FLAGS.episode_len
+        if self._steps_taken < FLAGS.episode_len - 1:
+            # Don't need to record the last action since there are no
+            # further decisions to be made at that point, so that
+            # information need never be presented to the model.
+            self._state[self._steps_taken][action] = 1
+        self._steps_taken += 1
+
+        # Update environment
+        _, reward, done, info = self._env.step(self._actions[action])
+        done = done or self._steps_taken == FLAGS.episode_len
+
+        return self._state, reward, done, info
+
+    def reset(self):
+        self._env.reset()
+        self._steps_taken = 0
+        self._state = np.zeros(
+            (FLAGS.episode_len - 1, len(self.action_space)), dtype=np.int32
+        )
+        return self._state
+
+    def close(self):
+        self._env.close()
+
+
+class Policy(nn.Module):
+    """A very simple actor critic policy model."""
+
+    def __init__(self):
+        super().__init__()
+        self.affine1 = nn.Linear(
+            (FLAGS.episode_len - 1) * len(FLAGS.flags), FLAGS.hidden_size
+        )
+        self.affine2 = nn.Linear(FLAGS.hidden_size, FLAGS.hidden_size)
+        self.affine3 = nn.Linear(FLAGS.hidden_size, FLAGS.hidden_size)
+        self.affine4 = nn.Linear(FLAGS.hidden_size, FLAGS.hidden_size)
+
+        # Actor's layer
+        self.action_head = nn.Linear(FLAGS.hidden_size, len(FLAGS.flags))
+
+        # Critic's layer
+        self.value_head = nn.Linear(FLAGS.hidden_size, 1)
+
+        # Action & reward buffer
+        self.saved_actions: List[SavedAction] = []
+        self.rewards: List[float] = []
+
+        # Keep exponential moving average of mean and standard
+        # deviation for use in normalization of the input.
+        self.moving_mean = MovingExponentialAverage(FLAGS.mean_smoothing)
+        self.moving_std = MovingExponentialAverage(FLAGS.std_smoothing)
+
+    def forward(self, x):
+        """Forward of both actor and critic"""
+        # Initial layer maps the sequence of one-hot vectors into a
+        # vector of the hidden size. Next layers stay with the same
+        # size and use residual connections.
+        x = F.relu(self.affine1(x))
+        x = x.add(F.relu(self.affine2(x)))
```

Comment:
i've mostly seen residual connections used when the branch has multiple layers. but if this works in practice, great :)

Reply:
I just figured I'd try it. I ran an experiment in the linked pastebin that shows that performance is substantially worse without the residual connections, so I left them in. (Though just removing the additional layers is also supported by that data)